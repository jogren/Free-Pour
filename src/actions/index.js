export const showSelectDrink = targetDrink => ({
  type: 'SHOW_SELECT_DRINK',
  targetDrink
});

export const hideSelectedDrink = () => ({
  type: 'HIDE_SELECTED_DRINK'
});

export const setCurrentCocktails = cocktails => ({
  type: 'SET_CURRENT_COCKTAILS',
  cocktails
});