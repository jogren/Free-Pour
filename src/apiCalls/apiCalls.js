export const fetchPopularCocktails = async () => {
  const url = 'https://www.thecocktaildb.com/api/json/v2/8673533/popular.php';
  try {
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error('There was an error fetch your data')
    }
    const popularCocktails = await response.json()
    return popularCocktails.drinks.map(drink => ({
      strDrink: drink.strDrink,
      strDrinkThumb: drink.strDrinkThumb,
    }))
  } catch(error) {
    throw new Error(error.message)
  }
}

export const fetchCocktailsByGenre = async (type) => {
  const url = `https://www.thecocktaildb.com/api/json/v1/1/filter.php?i=${type}`
  try {
    const response = await fetch(url)
    if(!response.ok) {
      throw new Error('There was an issue fetching your data by genre')
    }
    const cocktails = await response.json();
    return cocktails.drinks
  } catch (error) {
    throw new Error(error.message)

  }
}

export const fetchCocktailsBySearch = async (text) => {
  const url = `https://www.thecocktaildb.com/api/json/v1/1/search.php?s=${text}`
  try {
    const response = await fetch(url)
    if(!response.ok) {
      throw new Error('There was an issue fetching your data by name')
    }
    const cocktails = await response.json();
    return cocktails.drinks
  } catch(error) {
    throw new Error(error.message)
  }
}

export const fetchAllIngredients = async () => {
  const url = '  https://www.thecocktaildb.com/api/json/v1/1/list.php?i=list';
  try {
    const response = await fetch(url)
    if (!response.ok) {
      throw new Error('There was an issue fetching all ingredients')
    }
    const ingredients = await response.json();
    return ingredients.drinks.map(drink => drink.strIngredient1)
  } catch(error) {
    throw new Error(error.message)
  }
}

export const fetchMoreDrinkInfo = async (name) => {
  try {
    const response = await fetch(`https://www.thecocktaildb.com/api/json/v1/1/search.php?s=${name.split(' ').join('_')}`)
    if (!response.ok) {
      throw new Error('There was an error fetch your data')
    }
    const cocktail = await response.json();
    const cleanedCocktail = cocktail.drinks.map(drink => ({
      name: drink.strDrink,
      glass: drink.strGlass,
      instructions: drink.strInstructions,
      image: drink.strDrinkThumb,
      ingredients: [
        { measure: drink.strMeasure1, ingredient: drink.strIngredient1 },
        { measure: drink.strMeasure2, ingredient: drink.strIngredient2 },
        { measure: drink.strMeasure3, ingredient: drink.strIngredient3 },
        { measure: drink.strMeasure4, ingredient: drink.strIngredient4 },
        { measure: drink.strMeasure5, ingredient: drink.strIngredient5 },
        { measure: drink.strMeasure6, ingredient: drink.strIngredient6 }
      ]
    }))
    return cleanedCocktail[0]
  } catch(error) {
    throw new Error(error.message)
  }
}