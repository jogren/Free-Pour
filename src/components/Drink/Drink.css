.Drink_section {
  display: flex;
  justify-content: space-around;
  align-items: center;
  background-color: grey;
  margin: 10px;
  padding: 10px;
  position: relative;
}

.Drink_image {
  height: 100px;
}

.Drink_image-favorite {
  height: 50px;
  width: 45px;
  position: absolute;
  top: 5px;
  right: 5px;
}