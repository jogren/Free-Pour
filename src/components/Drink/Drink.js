import React, { Component } from 'react';
import { fetchMoreDrinkInfo } from '../../apiCalls/apiCalls';
import { connect } from 'react-redux';
import { showSelectDrink } from '../../actions';
import './Drink.css';

class Drink extends Component {
  constructor() {
    super();
    this.state = {
      selectedDrink: {}
    }

  }

  handleShowMore = async () => {
    const response = await fetchMoreDrinkInfo(this.props.name)
    this.props.showSelectDrink(response)
    // this.setState({ selectedDrink: response })
  }

  render() {
    const { image, name } = this.props
    return (
      <section className="Drink_section">
        <img className="Drink_image" src={image} alt={name}/>
        <h3>{name}</h3>
        <button onClick={this.handleShowMore}>How to Make</button>
      </section>
    )
  }
}

const mapStateToProps = ({ toggleSelectedDrink }) => ({
  toggleSelectedDrink,
});

const mapDispatchToProps = dispatch => ({
  showSelectDrink: (targetDrink) => dispatch(showSelectDrink(targetDrink))
});

export default connect(mapStateToProps, mapDispatchToProps)(Drink);