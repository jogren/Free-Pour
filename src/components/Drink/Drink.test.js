import React from 'react';
import { shallow } from 'enzyme';
import { Drink, mapStateToProps, mapDispatchToProps } from './Drink';
import { fetchMoreDrinkInfo } from '../../apiCalls/apiCalls';

import { showSelectDrink, toggleFavorite } from '../../actions';

jest.mock('../../apiCalls/apiCalls');

describe('Drink', () => {
  let wrapper;
  const mockShowSelectDrink = jest.fn();
  const mockToggleFavorite = jest.fn();
  const mockFavoriteCocktails = [
    {
      name: 'Moscow Mule',
      glass: 'Copper Mug',
      instructions: 'Combine ingredients...',
      image: 'image2.png',
      ingredients: [
        {
          guessed: false,
          ingredient: 'Vodka',
          measure: '2-3 oz'
        },
        {
          guessed: false,
          ingredient: 'Ginger Ale',
          measure: '5oz'
        },
        {
          guessed: false,
          ingredient: 'Lime Juice',
          measure: '2tbsp'
        }
      ]
    }
  ]

  beforeEach(() => {
    wrapper = shallow(<Drink 
      image='image.png'
      name='Mojito'
      showSelectDrink={mockShowSelectDrink}
      toggleFavorite={mockToggleFavorite}
      favoriteCocktails={mockFavoriteCocktails}
    />)
  })

  it('should take a snapshot when passed the correct data', () => {
    expect(wrapper).toMatchSnapshot();
  })

  describe('handleShowMore', () => {
    it('should invoke handleShowMore when the button is clicked', () => {
      wrapper.instance().handleShowMore = jest.fn();
      wrapper.instance().forceUpdate();

      wrapper.find('button').simulate('click');
      expect(wrapper.instance().handleShowMore).toHaveBeenCalled()
    })

    it('should invoke fetchMoreDrinkInfo and showSelectedDrink when handleShowMore is invoked', async () => {
      await wrapper.instance().handleShowMore();

      expect(fetchMoreDrinkInfo).toHaveBeenCalled();
      expect(mockShowSelectDrink).toHaveBeenCalled();
    })
  })

  describe('toggleFavorite', () => {
    it('should invoke toggleFavorite when the label is clicked', () => {
      wrapper.instance().toggleFavorite = jest.fn();
      wrapper.instance().forceUpdate();
    
      wrapper.find('label').simulate('click');
      expect(wrapper.instance().toggleFavorite).toHaveBeenCalled()
    })

    it('should invoke fetchMoreDrinkInfo with the correct drink name', () => {
      wrapper.instance().toggleFavorite({ name: 'Moscow Mule' })
      expect(fetchMoreDrinkInfo).toHaveBeenCalledWith('Moscow Mule')
    })
  })

  describe('toggleFavoriteLogic', () => {
    it('should filtered out drink if already present in cocktailFavorites', () => {
      const mockTargetCocktail = {
        id: 1,
        name: 'Moscow Mule'
      }
      wrapper.instance().toggleFavoriteLogic(mockTargetCocktail, true);

      expect(mockToggleFavorite).toHaveBeenCalledWith([])
    })

    it('should add drink if not already present in cocktailFavorites', () => {
      const mockTargetCocktail = {
        id: 1,
        name: 'Gin and Tonic'
      }
      wrapper.instance().toggleFavoriteLogic(mockTargetCocktail, false);

      expect(mockToggleFavorite).toHaveBeenCalledWith([...mockFavoriteCocktails, mockTargetCocktail])
    })
  })

  describe('mapStateToProps', () => {
    it('should return an object with the favoriteCocktails array', () => {
      const mockState = {
        selectedDrink: { id: 3, name: 'Gin and Tonic' },
        currentCocktails: [{ id: 1, name: 'Moscow Mule' }, { id: 2, name: 'Vodka Soda' }],
        favoriteCocktails: mockFavoriteCocktails
      };
      const expected = {
        favoriteCocktails: mockFavoriteCocktails
      };
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch with an showSelectDrink action', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = showSelectDrink({ id: 1, name: 'Moscow Mule' });
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.showSelectDrink({ id: 1, name: 'Moscow Mule' });
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch with an toggleFavorite action', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = toggleFavorite([{ id: 1, name: 'Moscow Mule' }, { id: 2, name: 'Vodka Soda' }]);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.toggleFavorite([{ id: 1, name: 'Moscow Mule' }, { id: 2, name: 'Vodka Soda' }])
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
})