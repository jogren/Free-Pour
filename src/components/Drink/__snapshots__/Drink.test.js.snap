// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drink should take a snapshot when passed the correct data 1`] = `
<section
  className="Drink_section"
>
  <img
    alt="Mojito"
    className="Drink_image"
    src="image.png"
  />
  <h3>
    Mojito
  </h3>
  <button
    onClick={[Function]}
  >
    See How to Make
  </button>
  <label
    onClick={[Function]}
  >
    <img
      alt="favorite this drink"
      className="Drink_image-favorite"
      src="favorite.png"
    />
    Click Here to Favorite!
  </label>
</section>
`;
